src/freertos_drivers/esp32/Esp32BootloaderHal.hxx

```diff
@@ -0,0 +1,562 @@
+            {
+                // start the OTA process at this point, if we have had a
+                // previous failure this will reset the OTA process so we can
+                // start fresh.
+                esp_err_t err = ESP_ERROR_CHECK_WITHOUT_ABORT(
+                    esp_ota_begin(esp_bl_state.target, OTA_SIZE_UNKNOWN,
+                                  &esp_bl_state.ota_handle));
+                should_abort = (err != ESP_OK);
+            }
+            else
+            {
+                LOG_ERROR("[Bootloader] Firmware does not appear to be valid "
+                          "or is for a different chip (%s - %x vs %s - %x).",
+                          CHIP_ID_NAMES[image_header->chip_id],
+                          image_header->chip_id,
+                          CHIP_ID_NAMES[esp_bl_state.chip_id],
+                          esp_bl_state.chip_id);
+                should_abort = true;
+            }
+        }
+        else
+        {
+            LOG_ERROR("[Bootloader] Image magic is incorrect: %d vs %d!",
+                      image_header->magic, ESP_IMAGE_HEADER_MAGIC);
+            should_abort = true;
+        }
+
+        // It would be ideal to abort the firmware upload at this point but the
+        // bootloader HAL does not offer a way to abort the transfer so instead
+        // reboot the node.
+        if (should_abort || esp_bl_state.ota_handle == 0)
+        {
+            // reset the RTC persistent variable to not enter bootloader on
+            // next restart.
+            bootloader_request = RTC_BOOL_FALSE;
+            esp_restart();
+        }
+    }
+    bootloader_led(LED_WRITING, true);
+    bootloader_led(LED_ACTIVE, false);
+    ESP_ERROR_CHECK(
+        esp_ota_write(esp_bl_state.ota_handle, data, size_bytes));
+    bootloader_led(LED_WRITING, false);
+    bootloader_led(LED_ACTIVE, true);
+}
+
+/// Callback from the bootloader to indicate that the full firmware file has
+/// been received.
+///
+/// @return zero if the firmware has been received and updated to be used for
+/// the next startup, otherwise non-zero and an error will be printed to the
+/// console.
+uint16_t flash_complete(void)
+{
+    LOG(INFO, "[Bootloader] Finalizing firmware update");
+    esp_err_t res =
+        ESP_ERROR_CHECK_WITHOUT_ABORT(esp_ota_end(esp_bl_state.ota_handle));
+    if (res == ESP_OK)
+    {
+        LOG(INFO,
+            "[Bootloader] Firmware appears valid, updating the next boot "
+            "partition to %s.", esp_bl_state.target->label);
+        res =
+            ESP_ERROR_CHECK_WITHOUT_ABORT(
+                esp_ota_set_boot_partition(esp_bl_state.target));
+        if (res != ESP_OK)
+        {
+            LOG_ERROR("[Bootloader] Failed to update the boot partition!");
+        }
+    }
+    else if (res == ESP_ERR_OTA_VALIDATE_FAILED)
+    {
+        LOG_ERROR("[Bootloader] Firmware image failed validation, aborting!");
+    }
+    return res != ESP_OK;
```

Comment:
I'll update to use `ERROR_FIRMWARE_CSUM` for failure cases but I can't use `ERROR_FIRMWARE_INCOMPATIBLE` since this is validated in `write_flash` (address zero) which doen't offer a clean abort sequence.

Should we consider adding an option for aborting inside `flush_flash_buffer`?

Reply:
https://github.com/bakerstu/openmrn/pull/719 covers the return value

Reply:
Logged https://github.com/bakerstu/openmrn/issues/721 to track an abort option for `write_flash`.